### PR TITLE
Edit frame component implementation

### PR DESCRIPTION
**Pull request by art-alexeyenko**
HUGE thanks to @mjarosch for the initial PR (https://github.com/Sitecore/jss/pull/1253) and doing the discovery, investigation and the component implementation.

This adds an Edit Frame component, similar to the one that was available for Sitecore's MVC site implementations.
The PR includes:
* Edit Frame component implementation for Angular, Vue, React (also usable with NextJS);
* Styleguide component showcase for Edit Frame for Angular, Vue, React, NextJS.

## Testing Details
Manual testing + new unit tests for each framework.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by sc-addypathania — CHANGES_REQUESTED** [review]
Awesome work. :star2:
See some minor comments below. :)

File: packages/create-sitecore-jss/src/templates/angular/src/app/components/styleguide-edit-frame/styleguide-edit-frame.component.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    this.editFrameProps = this.getEditFrameProps(this.rendering.dataSource);
+  }
+
+  getEditFrameProps(dataSource: string) {
+    return {
+      dataSource: dataSource
+        ? {
+            itemId: dataSource,
+            // databaseName: 'web',
+            // language: 'en', // optional params you can also set for datasource
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/create-sitecore-jss/src/templates/angular/src/app/components/styleguide-edit-frame/styleguide-edit-frame.component.ts:63]: yes, to describe optional parameters that can be used

File: packages/create-sitecore-jss/src/templates/react/src/components/Styleguide-EditFrame/index.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      Who framed Roger Rabbit? Hard to say. <br />
+      But JSS now allows to edit frame any piece of content on a page in editing mode. <br />
+      You can add web edit or field edit buttons, modify edit frame style through CSS class and
+      put the frame wherever you need it.
+      {props.children}
+    </EditFrame>
+  </StyleguideSpecimen>
+);
+
+const getEditFrameProps = (dataSource) => {
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/create-sitecore-jss/src/templates/react/src/components/Styleguide-EditFrame/index.js:17]: This is a javascript file - so we can't use types here.

File: packages/sitecore-jss-angular/src/components/editframe.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+  it('should render a frame when in pageEditing mode', () => {
+    editFrameComp.sitecore = editingContext;
+    fixture.detectChanges();
+
+    const frameDiv = de.query(By.css('div.scLooseFrameZone')).nativeElement;
+    const frameSpan = de.query(By.css('div.scLooseFrameZone > span.scChromeData')).nativeElement;
+    expect(frameDiv).toBeDefined();
+    expect(frameSpan).toBeDefined();
+    expect(frameSpan.textContent).toBe('{}');
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-angular/src/components/editframe.component.spec.ts:114]: The tests here only compare strings or booleans - either toBe or toEqual should be fine for this.

File: packages/sitecore-jss-react/src/components/EditFrame.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+export interface EditFrameProps {
+  dataSource?: EditFrameDataSource;
+  buttons?: (FieldEditButton | WebEditButton | '|')[];
+  title?: string;
+  tooltip?: string;
+  cssClass?: string;
+  parameters?: Record<string, string | number | boolean | undefined | null>;
+}
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-react/src/components/EditFrame.tsx:77]: It is indeed. I'll remove this.

File: packages/sitecore-jss-angular/src/components/editframe.component.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  chromeData = '';
+
+  ngOnChanges() {
+    this.isEditing = this.sitecore.context.pageEditing || false;
+    if (!this.isEditing) {
+      return;
+    }
+
+    const chromeData: Record<string, unknown> = {
+      displayName: this.title,
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-angular/src/components/editframe.component.ts:57]: Makes sense

File: packages/sitecore-jss-angular/src/components/editframe.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    fixture = TestBed.createComponent(TestComponent);
+    fixture.detectChanges();
+
+    de = fixture.debugElement.query(By.css('sc-edit-frame'));
+    editFrameComp = fixture.componentInstance;
+    editFrameComp.sitecore = normalContext;
+  });
+
+  it('should render wrapped element when not in pageEditing mode', () => {
+    editFrameComp.sitecore = normalContext;
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-angular/src/components/editframe.component.spec.ts:99]: I removed the assignment in `beforeEach` . I think having context defined explicitly in each test describes the use/testcase better.

File: packages/sitecore-jss-angular/src/components/editframe.component.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import {
+  EditFrameDataSource,
+  ChromeCommand,
+  EditButtonTypes,
+  mapButtonToCommand,
+} from '@sitecore-jss/sitecore-jss/utils';
+import { LayoutServiceContextData, RouteData } from '@sitecore-jss/sitecore-jss/layout';
+
+@Component({
+  selector: 'sc-edit-frame,[sc-edit-frame]',
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-angular/src/components/editframe.component.ts:11]: Is there a reason to? Styleguide components will give some guidance on how to use the edit frame.

File: packages/sitecore-jss-vue/src/components/EditFrame.test.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    const rendered = mount(EditFrameWithText, {
+      props: {
+        context: editingContext,
+      },
+    });
+    expect(rendered.find('div.scLooseFrameZone').text()).toContain('Test text');
+  });
+});
+
+it('should render the title and tooltip', () => {
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply packages/sitecore-jss-vue/src/components/EditFrame.test.ts]: indentation issue, fixed.

**Review by sc-addypathania — APPROVED** [review]
Looks good. :rocket:

**Pull request merged**